==next3_snapshot_debug.patch==

next3: snapshot debugging support

Control snapshot debug level via debugfs entry /next3/snapshot-debug.
Control snapshot unit tests via debugfs entries /next3/test-XXX.

Signed-off-by: Amir Goldstein <amir73il@users.sf.net>

--------------------------------------------------------------------------------
diff -Nuarp a/fs/next3/Kconfig b/fs/next3/Kconfig
--- a/fs/next3/Kconfig	2010-07-26 09:22:09.585098571 +0300
+++ b/fs/next3/Kconfig	2010-07-26 09:22:09.105096009 +0300
@@ -86,3 +86,11 @@ config NEXT3_FS_SECURITY
 
 	  If you are not using a security module that requires using
 	  extended attributes for file security labels, say N.
+
+config NEXT3_FS_DEBUG
+	bool "snapshot debugging support"
+	depends on NEXT3_FS && DEBUG_FS
+	default y
+	help
+	  Control snapshot debug level via debugfs entry /next3/snapshot-debug.
+	  Control snapshot unit tests via debugfs entries /next3/test-XXX.
diff -Nuarp a/fs/next3/Makefile b/fs/next3/Makefile
--- a/fs/next3/Makefile	2010-07-26 09:22:09.595106855 +0300
+++ b/fs/next3/Makefile	2010-07-26 09:22:09.105096009 +0300
@@ -18,3 +18,4 @@ next3-y	+= snapshot.o snapshot_ctl.o buf
 next3-$(CONFIG_NEXT3_FS_XATTR)	 += xattr.o xattr_user.o xattr_trusted.o
 next3-$(CONFIG_NEXT3_FS_POSIX_ACL) += acl.o
 next3-$(CONFIG_NEXT3_FS_SECURITY)	 += xattr_security.o
+next3-$(CONFIG_NEXT3_FS_DEBUG)	 += snapshot_debug.o
diff -Nuarp a/fs/next3/next3_jbd.h b/fs/next3/next3_jbd.h
--- a/fs/next3/next3_jbd.h	2010-07-26 09:22:09.685099645 +0300
+++ b/fs/next3/next3_jbd.h	2010-07-26 09:22:09.195025180 +0300
@@ -21,6 +21,7 @@
 #include <linux/fs.h>
 #include <linux/jbd.h>
 #include "next3.h"
+#include "snapshot_debug.h"
 
 #define NEXT3_JOURNAL(inode)	(NEXT3_SB((inode)->i_sb)->s_journal)
 
diff -Nuarp a/fs/next3/snapshot_debug.c b/fs/next3/snapshot_debug.c
--- a/fs/next3/snapshot_debug.c	2010-07-26 09:22:09.725105367 +0300
+++ b/fs/next3/snapshot_debug.c	2010-07-26 09:22:09.235071055 +0300
@@ -0,0 +1,104 @@
+/*
+ * linux/fs/next3/snapshot_debug.c
+ *
+ * Written by Amir Goldstein <amir73il@users.sf.net>, 2008
+ *
+ * Copyright (C) 2008-2010 CTERA Networks
+ *
+ * This file is part of the Linux kernel and is made available under
+ * the terms of the GNU General Public License, version 2, or at your
+ * option, any later version, incorporated herein by reference.
+ *
+ * Next3 snapshot debugging.
+ */
+
+
+#include <linux/module.h>
+#include <linux/proc_fs.h>
+#include <linux/debugfs.h>
+#include "snapshot.h"
+
+/*
+ * debugfs tunables
+ */
+
+const char *snapshot_indent = SNAPSHOT_INDENT_STR + SNAPSHOT_INDENT_MAX;
+
+/*
+ * Tunable delay values per snapshot operation for testing of
+ * COW race conditions and master snapshot_mutex lock
+ */
+static const char *snapshot_test_names[SNAPSHOT_TESTS_NUM] = {
+	/* delay completion of snapshot create|take */
+	"test-take-delay-msec",
+	/* delay completion of snapshot shrink|cleanup */
+	"test-delete-delay-msec",
+	/* delay completion of COW operation */
+	"test-cow-delay-msec",
+	/* delay submission of tracked read */
+	"test-read-delay-msec",
+	/* delay completion of COW bitmap operation */
+	"test-bitmap-delay-msec",
+};
+
+#define SNAPSHOT_TEST_NAMES (sizeof(snapshot_test_names) / \
+			     sizeof(snapshot_test_names[0]))
+
+u16 snapshot_enable_test[SNAPSHOT_TESTS_NUM] __read_mostly = {0};
+u8 snapshot_enable_debug __read_mostly = 1;
+
+static struct dentry *next3_debugfs_dir;
+static struct dentry *snapshot_debug;
+static struct dentry *snapshot_version;
+static struct dentry *snapshot_test[SNAPSHOT_TESTS_NUM];
+
+static char snapshot_version_str[] = NEXT3_SNAPSHOT_VERSION;
+struct debugfs_blob_wrapper snapshot_version_blob = {
+	.data = snapshot_version_str,
+	.size = sizeof(snapshot_version_str)
+};
+
+/*
+ * next3_create_debugfs_entry - register next3 debug hooks
+ * Void function doesn't return error if debug hooks are not registered.
+ */
+void next3_create_debugfs_entry(void)
+{
+	int i;
+
+	next3_debugfs_dir = debugfs_create_dir("next3", NULL);
+	if (!next3_debugfs_dir)
+		return;
+	snapshot_debug = debugfs_create_u8("snapshot-debug", S_IRUGO|S_IWUSR,
+					   next3_debugfs_dir,
+					   &snapshot_enable_debug);
+	snapshot_version = debugfs_create_blob("snapshot-version", S_IRUGO,
+					       next3_debugfs_dir,
+					       &snapshot_version_blob);
+	for (i = 0; i < SNAPSHOT_TESTS_NUM && i < SNAPSHOT_TEST_NAMES; i++)
+		snapshot_test[i] = debugfs_create_u16(snapshot_test_names[i],
+					      S_IRUGO|S_IWUSR,
+					      next3_debugfs_dir,
+					      &snapshot_enable_test[i]);
+}
+
+/*
+ * next3_remove_debugfs_entry - unregister next3 debug hooks
+ * checks if the hooks have been registered before unregistering them.
+ */
+void next3_remove_debugfs_entry(void)
+{
+	int i;
+
+	if (!next3_debugfs_dir)
+		return;
+	for (i = 0; i < SNAPSHOT_TESTS_NUM && i < SNAPSHOT_TEST_NAMES; i++)
+		if (snapshot_test[i])
+			debugfs_remove(snapshot_test[i]);
+	if (snapshot_version)
+		debugfs_remove(snapshot_version);
+	if (snapshot_debug)
+		debugfs_remove(snapshot_debug);
+	debugfs_remove(next3_debugfs_dir);
+}
+
diff -Nuarp a/fs/next3/snapshot_debug.h b/fs/next3/snapshot_debug.h
--- a/fs/next3/snapshot_debug.h	2010-07-26 09:22:09.725105367 +0300
+++ b/fs/next3/snapshot_debug.h	2010-07-26 09:22:09.245068260 +0300
@@ -0,0 +1,105 @@
+/*
+ * linux/fs/next3/snapshot_debug.h
+ *
+ * Written by Amir Goldstein <amir73il@users.sf.net>, 2008
+ *
+ * Copyright (C) 2008-2010 CTERA Networks
+ *
+ * This file is part of the Linux kernel and is made available under
+ * the terms of the GNU General Public License, version 2, or at your
+ * option, any later version, incorporated herein by reference.
+ *
+ * Next3 snapshot debugging.
+ */
+
+#ifndef _LINUX_NEXT3_SNAPSHOT_DEBUG_H
+#define _LINUX_NEXT3_SNAPSHOT_DEBUG_H
+
+#ifdef CONFIG_NEXT3_FS_DEBUG
+#include <linux/delay.h>
+
+#define SNAPSHOT_INDENT_MAX 4
+#define SNAPSHOT_INDENT_STR "\t\t\t\t"
+#define KERN_LEVEL_STR "<%d>"
+#define SNAP_KERN_LEVEL(n) ((n)+2) /* 1 = KERN_ERR, ..., 5 = KERN_DEBUG */
+
+#define SNAPTEST_TAKE	0
+#define SNAPTEST_DELETE	1
+#define SNAPTEST_COW	2
+#define SNAPTEST_READ	3
+#define SNAPTEST_BITMAP	4
+#define SNAPSHOT_TESTS_NUM	5
+
+extern const char *snapshot_indent;
+extern u8 snapshot_enable_debug;
+extern u16 snapshot_enable_test[SNAPSHOT_TESTS_NUM];
+extern u8 cow_cache_offset;
+
+#define snapshot_test_delay(i)		     \
+	do {							       \
+		if (snapshot_enable_test[i])			       \
+			msleep_interruptible(snapshot_enable_test[i]); \
+	} while (0)
+
+#define snapshot_test_delay_per_ticks(i, n)	    \
+	do {								\
+		if (snapshot_enable_test[i] && (n))			\
+			msleep_interruptible(				\
+				(snapshot_enable_test[i]/(n))+1);	\
+	} while (0)
+
+#define snapshot_debug_l(n, l, f, a...)					\
+	do {								\
+		if ((n) <= snapshot_enable_debug &&			\
+		    (l) <= SNAPSHOT_INDENT_MAX) {			\
+			printk(KERN_LEVEL_STR "snapshot: %s" f,		\
+				   SNAP_KERN_LEVEL(n),			\
+			       snapshot_indent - (l),			\
+				   ## a);				\
+		}							\
+	} while (0)
+
+#define snapshot_debug(n, f, a...)	snapshot_debug_l(n, 0, f, ## a)
+
+#define SNAPSHOT_DEBUG_ONCE int once = 1
+#define snapshot_debug_once(n, f, a...)					\
+	do {								\
+		if (once) {						\
+			snapshot_debug(n, f, ## a);			\
+			once = 0;					\
+		}							\
+	} while (0)
+
+extern void next3_create_debugfs_entry(void);
+extern void next3_remove_debugfs_entry(void);
+
+static inline void init_next3_snapshot_debug(void)
+{
+	next3_create_debugfs_entry();
+}
+
+static inline void exit_next3_snapshot_debug(void)
+{
+	next3_remove_debugfs_entry();
+}
+
+#else
+#define snapshot_test_delay(i)
+#define snapshot_test_delay_per_ticks(i, n)
+#define snapshot_debug(n, f, a...)
+#define snapshot_debug_l(n, l, f, a...)
+#define snapshot_debug_once(n, f, a...)
+#define SNAPSHOT_DEBUG_ONCE
+#define init_next3_snapshot_debug()
+#define exit_next3_snapshot_debug()
+#endif
+
+
+/* debug levels */
+#define SNAP_ERR	1 /* errors and summary */
+#define SNAP_WARN	2 /* warnings */
+#define SNAP_INFO	3 /* info */
+#define SNAP_DEBUG	4 /* debug */
+#define SNAP_DUMP	5 /* dump snapshot file */
+
+#endif	/* _LINUX_NEXT3_SNAPSHOT_DEBUG_H */
diff -Nuarp a/fs/next3/snapshot.h b/fs/next3/snapshot.h
--- a/fs/next3/snapshot.h	2010-07-26 09:22:09.715036670 +0300
+++ b/fs/next3/snapshot.h	2010-07-26 09:22:09.225092179 +0300
@@ -17,6 +17,7 @@
 
 #include <linux/delay.h>
 #include "next3_jbd.h"
+#include "snapshot_debug.h"
 
 
 #define NEXT3_SNAPSHOT_VERSION "next3 snapshot v1.0.12 (25-Jul-2010)"
@@ -107,11 +108,13 @@ extern void next3_snapshot_destroy(struc
 
 static inline int init_next3_snapshot(void)
 {
+	init_next3_snapshot_debug();
 	return 0;
 }
 
 static inline void exit_next3_snapshot(void)
 {
+	exit_next3_snapshot_debug();
 }
 
 
